tf-dataflow/tfproc/tfproc.py

```diff
@@ -0,0 +1,213 @@
+    colorspace = b"RGB"
+    channels = 3
+    image_format = b"PNG"
+    xmins, xmaxes, ymins, ymaxes = bboxes
+
+    example = tf.train.Example(
+        features=tf.train.Features(
+            feature={
+                "image/height": _int64_feature(height),
+                "image/width": _int64_feature(width),
+                "image/colorspace": _bytes_feature(colorspace),
+                "image/channels": _int64_feature(channels),
+                "image/object/class/label": _int64_list_feature(
+                    labels
+                ),  # model expects 1-based
+                "image/object/bbox/xmin": _float_list_feature(xmins),
+                "image/object/bbox/xmax": _float_list_feature(xmaxes),
+                "image/object/bbox/ymin": _float_list_feature(ymins),
+                "image/object/bbox/ymax": _float_list_feature(ymaxes),
+                "image/format": _bytes_feature(image_format),
+                "image/filename": _bytes_feature(
+                    bytes(os.path.basename(filename), "utf-8")
+                ),
+                "image/encoded": _bytes_feature(image_buffer),
+            }
+        )
+    )
+    return example
+
+
+def _bytes_feature(value):
+    """Returns a bytes_list from a string / byte."""
+    if isinstance(value, type(tf.constant(0))):
+        value = (
+            value.numpy()
+        )  # BytesList won't unpack a string from an EagerTensor.
+    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))
+
+
+def _float_feature(value):
+    return _float_list_feature([value])
+
+
+def _float_list_feature(value):
+    """Returns a float_list from a float / double."""
+    return tf.train.Feature(float_list=tf.train.FloatList(value=value))
+
+
+def _int64_feature(value):
+    """Returns an int64_list from a bool / enum / int / uint."""
+    return _int64_list_feature([value])
+
+
+def _int64_list_feature(value):
+    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))
+
+
+class AppParamLister(PTransform):
+    def __init__(self, execution_source):
+        super().__init__()
+        self.execution_source = execution_source
+
+    def expand(self, pcollection):
+        app_params_pattern = [
+            os.path.join(self.execution_source, "urn:app_params:*/*")
+        ]
+        logging.info("App params pattern %s", app_params_pattern)
+
+        app_params_coll = [f for f in tf.io.gfile.glob(app_params_pattern[0])]
+        logging.info(
+            "App param collection %s, length %d",
+            app_params_coll,
+            len(app_params_coll),
+        )
+
+        return pcollection | "Create instances" >> beam.Create(app_params_coll)
+
+
+class ProcessInstance(beam.DoFn):
+    def __init__(self, output_dir, eval_pct, label=None):
+        super().__init__(label)
+        self.output_dir = output_dir
+        self.eval_pct = eval_pct
+
+    def process(self, element, *args, **kwargs):
```

Comment:
Suggest breaking this down to smaller instance methods to improve readability and maintainability.

Reply:
Done.